File: MobMuPlatEditor-Java/src/com/iglesiaintermedia/MobMuPlatEditor/controls/MMPButton.java
package com.iglesiaintermedia.MobMuPlatEditor.controls;

import java.awt.Color;
import java.awt.Rectangle;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.ArrayList;

import javax.swing.BorderFactory;
import javax.swing.border.Border;

public class MMPButton extends MMPControl implements MouseListener{
	static final int EDGE_RADIUS = 5;
	int value;
	RoundedPanel buttonPanel;
	
	public MMPButton(MMPButton otherButton){//copy constructor
		this(otherButton.getBounds());//normal constructor
		this.setColor(otherButton.color);
		this.setHighlightColor(otherButton.highlightColor);
		this.address=otherButton.address;
		
	}
	
	public MMPButton(Rectangle frame){
		super();
		address = "/myButton";
		buttonPanel = new RoundedPanel();
		buttonPanel.setCornerRadius(EDGE_RADIUS);
		add(buttonPanel);
		
		this.addMouseListener(this);
		this.addMouseMotionListener(this);
		this.setColor(this.color);
		this.setBounds(frame);
	}
	
	public void setBounds(Rectangle frame){
		super.setBounds(frame);
	    buttonPanel.setBounds(0,0,this.getWidth(), this.getHeight());
	}
	
	public void setColor(Color newColor){
		super.setColor(newColor);
		
		buttonPanel.setBackground(newColor);
		
	}
	
	public void setValue(int inValue){
		value=inValue;
		
		Object[] args = new Object[]{new Integer(value)};
		editingDelegate.sendMessage(address, args);
		
	}
	
	
	
	public void mousePressed(MouseEvent e) {
		//System.out.print(e.getClickCount() + " click(s)");
		super.mousePressed(e);
		
		if(!editingDelegate.isEditing()){
	      setValue(1);
	      buttonPanel.setBackground(highlightColor);
	      
	    }
	}


	@Override
	public void mouseReleased(MouseEvent e) {
		super.mouseReleased(e);
	    if(!editingDelegate.isEditing()){
	    	setValue(0);
	    	buttonPanel.setBackground(color);
	    }
		
	}
	
	//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object
	//for button, any message means a instantaneous touch down and touch up
	//it does not respond to "set" anything
	public void receiveList(ArrayList<Object> messageArray){
		if (messageArray.size()>0 && (messageArray.get(0) instanceof Float || messageArray.get(0) instanceof Integer) ){
	       setValue(1);
	       setValue(0);
	    }
	}
	
	
}

File: MobMuPlatEditor-OSXLionAndUp/MobMuPlatEditor/Controls/MMPButton.m
//
//  MMPButton.m
//  MobMuPlatEd1
//
//  Created by Daniel Iglesia on 12/28/12.
//  Copyright (c) 2012 Daniel Iglesia. All rights reserved.
//

#import "MMPButton.h"
#define EDGE_RADIUS 5

@implementation MMPButton

- (id)initWithFrame:(NSRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        self.address=@"/myButton";
      self.layer.backgroundColor=self.color.CGColor;
        self.layer.cornerRadius=EDGE_RADIUS;
        [self addHandles];
      [self resizeSubviewsWithOldSize:self.frame.size];
    }
    return self;
}

-(void)hackRefresh{
    [super hackRefresh];
    self.layer.cornerRadius=EDGE_RADIUS;
}

-(void)mouseDown:(NSEvent *)event{
    [super mouseDown:event];
    if(![self.editingDelegate isEditing] && self.enabled){
        [self setValue:1];
    }
}


-(void)mouseUp:(NSEvent *)event{
    [super mouseUp:event];
    if(![self.editingDelegate isEditing] && self.enabled)
        [self setValue:0];
}

-(void)setColor:(NSColor *)color{
    [super setColor:color];
  self.layer.backgroundColor=color.CGColor;
}

-(void)setValue:(int)inVal{
	_value=inVal;
    
  if(self.value==1)self.layer.backgroundColor=self.highlightColor.CGColor;
  else self.layer.backgroundColor=self.color.CGColor;
	 
	//send out message
    NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
    [formattedMessageArray addObject:self.address];
    [formattedMessageArray addObject:[NSNumber numberWithInt:self.value]];
    [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
}

//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object
//for button, any message means a instantaneous touch down and touch up
//it does not respond to "set" anything
-(void)receiveList:(NSArray *)inArray{
  [super receiveList:inArray];
    if ([inArray count]>0 && [[inArray objectAtIndex:0] isKindOfClass:[NSNumber class]]){
        //[self setValue:(int)[(NSNumber*)[inArray objectAtIndex:0] floatValue]];
        [self setValue:1];
        [self setValue:0];
    }
}





@end
